core/src/main/java/hudson/util/RingBufferLogHandler.java

```diff
@@ -86,13 +75,36 @@ public static int getDefaultRingBufferSize() {
     }
 
     @Override
-    public synchronized void publish(LogRecord record) {
-        int len = records.length;
-        records[(start + size) % len] = new LogRecordRef(record);
-        if (size == len) {
-            start = (start + 1) % len;
-        } else {
-            size++;
+    public void publish(LogRecord record) {
+        if (record == null) {
+            return;
+        }
+        if (JenkinsJVM.isJenkinsJVM() && record.getParameters() != null) {
+            try {
+                LogRecord clone = new LogRecord(record.getLevel(), dummyFormatter.formatMessage(record));
+                clone.setLoggerName(record.getLoggerName());
+                clone.setMillis(record.getMillis());
+                clone.setSequenceNumber(record.getSequenceNumber());
+                clone.setSourceClassName(record.getSourceClassName());
+                clone.setSourceMethodName(record.getSourceMethodName());
+                clone.setThreadID(record.getThreadID());
+                Throwable t = record.getThrown();
+                if (t != null) {
+                    clone.setThrown(new ProxyException(t));
+                }
+                record = clone;
+            } catch (Exception e) {
+                e.printStackTrace();
```

Comment:
should this be using an actual logger?

Reply:
No, because that could cause a cycle of sorts. In general you have to be cautious about what you do in logger methods.

Reply:
> In general you have to be cautious

(advice I should have taken more to heart)